llsd/base.py

```diff
@@ -399,53 +395,71 @@ def _getc(self, num=1):
+        buff = self._buffer
+        read_idx = self._index
+        cc = 0
         while True:
-            cc = self._getc()
-            if found_escape:
-                if found_hex:
-                    if found_digit:
-                        found_escape = False
-                        found_hex = False
-                        found_digit = False
-                        byte <<= 4
-                        byte |= _hex_as_nybble(cc)
-                        parts.append(byte)
-                        byte = 0
+            try:
+                cc = buff[read_idx]
+                read_idx += 1
+
+                if cc == _BACKSLASH_ORD:
+                    # Backslash, figure out if this is an \xNN hex escape or
+                    # something like \t
+                    cc = buff[read_idx]
+                    read_idx += 1
+                    if cc == _X_ORD:
+                        # It's a hex escape. char is the value of the two
+                        # following hex nybbles
+                        cc = int(chr(buff[read_idx]), 16) << 4
+                        read_idx += 1
+                        cc |= int(chr(buff[read_idx]), 16)
+                        read_idx += 1
                     else:
-                        found_digit = True
-                        byte = _hex_as_nybble(cc)
-                elif cc == b'x':
-                    found_hex = True
-                else:
-                    found_escape = False
-                    # escape char preceding anything other than the chars in
-                    # _escaped just results in that same char without the
-                    # escape char
-                    parts.extend(self._escaped.get(cc, cc))
-            elif cc == b'\\':
-                found_escape = True
-            elif cc == delim:
-                break
-            else:
-                parts.extend(cc)
+                        # escape char preceding anything other than the chars
+                        # in _escaped just results in that same char without
+                        # the escape char
+                        cc = self._escaped.get(cc, cc)
+                elif cc == delim_ord:
+                    break
+            except IndexError:
+                # We can be reasonably sure that any IndexErrors inside here
+                # were caused by an out-of-bounds `buff[read_idx]`.
+                self._index = read_idx
+                self._error("Trying to read past end of buffer")
+
+            decode_buff[insert_idx] = cc
+            insert_idx += 1
+
+            # We inserted a character, check if we need to expand the buffer.
+            if insert_idx % _DECODE_BUFF_ALLOC_SIZE == 0:
+                # Any additions may now overflow the decoding buffer, make
+                # a new expanded buffer containing the existing contents.
+                decode_buff = bytearray(decode_buff)
+                decode_buff.extend(b"\x00" * _DECODE_BUFF_ALLOC_SIZE)
```

Comment:
Would it be faster still to do something like:

    self._decode_buffs = [bytearray(_DECODE_BUFF_ALLOC_SIZE)]
    # ...
    buff_idx = 0
    decode_buff = self._decode_buffs[buff_idx]
    # ...
    try:
        decode_buff[insert_idx] = cc
    except IndexError:
        buff_idx += 1
        try:
            decode_buff = self._decode_buffs[buff_idx]
        except IndexError:
            decode_buff = bytearray(_DECODE_BUFF_ALLOC_SIZE)
            self._decode_buffs.append(decode_buff)
        insert_idx = 0
        decode_buff[insert_idx] = cc
    insert_idx += 1
    # ...
    try:
        return b''.join(self._decode_buffs[:-1] + [decode_buff[:insert_idx]]).decode('utf-8')
    except UnicodeDecodeError as exc:
        # ...

Reply:
I guess it depends on how many oversized strings you expect in the input stream. The code you submitted leaves you with a single larger buffer, prepared to handle many oversized strings without further allocations. My suggestion above should expand faster the first time, but requires consolidating multiple `bytearray`s for each oversized string. Plus which it's admittedly more complex.

I still suggest the code you wrote would be improved by catching `IndexError` on `decode_buff` assignment, though, rather than testing `insert_idx` as shown. Also I suspect there's a bug in your test.

Say you're working on the second oversized string in the same input stream, so `len(self._decode_buff) == 2*_DECODE_BUFF_ALLOC_SIZE`. You just inserted the byte at `(_DECODE_BUFF_ALLOC_SIZE - 1)` and incremented `insert_idx`. Now, although `insert_idx == len(decode_buff)/2`, `insert_idx % _DECODE_BUFF_ALLOC_SIZE` in fact equals 0, so you extend `decode_buff` -- resulting in `_DECODE_BUFF_ALLOC_SIZE` bytes of garbage in the buffer.

Reply:
> The code you submitted leaves you with a single larger buffer, prepared to handle many oversized strings without further allocations.

Whoops, not true: your expanded `decode_buff` isn't stored in `self._decode_buff`, so subsequent oversized strings would in fact require a new expanded `decode_buff`.

Reply:
>Whoops, not true: your expanded decode_buff isn't stored in self._decode_buff, so subsequent oversized strings would in fact require a new expanded decode_buff.

Yep, my reasoning there was I sometimes use long-lived `LLSD*Parser` objects and I was concerned about keeping a large scratch buffer around just because I'd previously parsed a large string.

> `except IndexError:`

That is a good point! I forgot that in Python it's much less expensive to handle the (potential) `IndexError` than check if we're about to overflow. The internal `IndexError` check is priced in and you don't really get to opt out anyway. Doing an opportunistic assignment and doing my buffer copy in the `except IndexError` shaves off about 100ms! I think this strategy ends up being easier to read as well (I didn't feel great about the `%`.)

The buffer copy itself isn't terribly expensive relative to juggling + concatenating multiple buffers though. Buffer concatenation was a tiny bit slower for payloads containing strings mostly under 1024 bytes.